docs(open-payments): Add readme docs (#1207)
interledger/rafiki: packages/open-payments/README.md

1. Create an Open Payments client

In this case, since Amazon Marketplace wants to make requests that require authorization, it will need to create an `AuthenticatedClient`:

```ts
import { createAuthenticatedClient } from '@interledger/open-payments'

const client = await createAuthenticatedClient({
  paymentPointerUrl: 'https://amazon.com/usa',
  keyId: KEY_ID,
  privateKey: PRIVATE_KEY
  // The public JWK with this key (and keyId) would be available at https://amazon.com/usa/jwks.json
})
```

2.  Get `PaymentPointers`

Grab the payment pointers of the parties:

```ts
const shoeShopPaymentPointer = await client.paymentPointer.get({
  url: 'https://happy-life-bank/shoe-shop'
})

const customerPaymentPointer = await client.paymentPointer.get({
  url: 'https://cloud-nine-wallet/alice'
})
```

3. Create `IncomingPayment`

Amazon's backend gets a grant to create an `IncomingPayment` on the merchant's wallet:

```ts
const incomingPaymentGrant = await client.grant.request(
  { url: shoeShopPaymentPointer.authServer },
  {
    access_token: {
      access: [
        {
          type: 'incoming-payment',
          actions: ['read-all', 'create']
        }
      ]
    }
  }
)

> 

Reviewer: I'd close the `ts` code block here, write somthing like 
"and creates an incoming payment using the access token from the grant"
and then I'd start the new `ts` code block with the incoming payment request.

const incomingPayment = await client.incomingPayment.create(
  {
    paymentPointer: shoeShopPaymentPointer.authServer,
    accessToken: incomingPaymentGrant.access_token.value
  },
  {
    incomingAmount: {
      assetCode: 'USD',
      assetScale: 2,
      value: '5000'
    },
    description: 'Purchase at Shoe Shop',
    externalRef: '#INV2022-8363828'
  }
)
```

4. Create `Quote`

> Then, it'll get a grant to create a `Quote` on Alice's payment pointer, which will give the amount it'll cost Alice to make the payment (with the ILP fees + her wallet's fees)

Reviewer: If we don't change it, quote grant requests are currently also interactive. The way quoting works right now, I think we should be able to make it non-interactive. But that may not be the case in the future with multiple payment methods.

Author: I can make an issue to have quote grant interactivity behind a flag (similar to `INCOMING_PAYMENT_INTERACTION`) and disable it by default

Author: - #1215

Reviewer: Do you want to get #1215 in first before merging this or do you want to mention that quotes are interactive right now and get this in first?

Author: I've looked, it doesn't seem to be too big of a lift so I'll try to get that in first

# Open Payments

[Open Payments](../../docs/open-payments.md) is an API standard that allows third-parties (with the account holder's consent) to initiate payments and to view the transaction history on the account holder's account.

Open Payments consists of two OpenAPI specifications, a **resource server** which exposes APIs for performing functions against the underlying accounts and an **authorization server** which exposes APIs compliant with the [GNAP](../../docs/glossary.md#grant-negotiation-authorization-protocol) standard for getting grants to access the resource server APIs.

This package provides TypeScript & NodeJS tools for using Open Payments:

> - Exported TypeScript types generated directly from the Open Payments Open API specifications

Reviewer 2: ```suggestion
- Exported TypeScript types generated directly from the Open Payments OpenAPI specifications
```

| ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `paymentPointerUrl` | The valid payment pointer with which the client making requests will identify itself. A JSON Web Key Set document that includes the public key that the client instance will use to protect requests MUST be available at the `{paymentPointerUrl}/jwks.json` url. This will be used as the `client` field during [Grant Creation](https://docs.openpayments.guide/reference/post-request). |

> | `privateKey`        | The the private EdDSA-Ed25519 key bound to the payment pointer, and used to sign the authenticated requests with. As mentioned above, a public JWK document signed with this key MUST be available at the `{paymentPointerUrl}/jwks.json` url.                                                                                                                                              |

Reviewer 2: ```suggestion
| `privateKey`        | The private EdDSA-Ed25519 key bound to the payment pointer, and used to sign the authenticated requests with. As mentioned above, a public JWK document signed with this key MUST be available at the `{paymentPointerUrl}/jwks.json` url.                                                                                                                                              |
```

```ts
const quoteGrant = await client.grant.request(
  { url: alicePaymentPointer.authServer },
  {
    access_token: {
      access: [
        {
          type: 'quote',
          actions: ['create', 'read']
        }
      ]
    }
  }
)

const quote = await client.quote.create(
  {
    paymentPointer: customerPaymentPointer.id,
    accessToken: quoteGrant.access_token.value
  },
  { receiver: incomingPayment.id }
)

// quote.sendAmount.value = '5200'
```

5. Create `OutgoingPayment` grant & start interaction flow:

The final step for Amazon's backend system will be to create an `OutgoingPayment` on Alice's wallet. Before this, however, Amazon will need to create an outgoing payment grant, which typically requires some sort of interaction with Alice. Amazon will need to facilitate this interaction with Alice (e.g. redirect her to a webpage with a dialog) to get her consent for creating an `OutgoingPayment` on her account. The detailed sequence for how this is achieved can be found [here](../../docs/grant-interaction.md).

> 7. Create `OutgoingPayment`:

Reviewer 2: ```suggestion
6. Create `OutgoingPayment`:
```

# Open Payments

[Open Payments](../../docs/open-payments.md) is an API standard that allows third-parties (with the account holder's consent) to initiate payments and to view the transaction history on the account holder's account.

> Open Payments consists of two Open API specifications, a **resource server** which exposes APIs for performing functions against the underlying accounts and an **authorization server** which exposes APIs compliant with the [GNAP](../../docs/glossary.md#grant-negotiation-authorization-protocol) standard for getting grants to access the resource server APIs.

Reviewer 2: ```suggestion
Open Payments consists of two OpenAPI specifications, a **resource server** which exposes APIs for performing functions against the underlying accounts and an **authorization server** which exposes APIs compliant with the [GNAP](../../docs/glossary.md#grant-negotiation-authorization-protocol) standard for getting grants to access the resource server APIs.
```

This package provides TypeScript & NodeJS tools for using Open Payments:

- Exported TypeScript types generated directly from the Open Payments Open API specifications
- A NodeJS client that exposes Open Payment APIs:
  - Signs requests with provided key
  - Validates responses against Open API specs

## Who is this package for?

This package could be used by (but not limited to):

- eCommerce backend services to facilitate payments between their wallet and a customer's wallet (example below)
- An app to display an account holder's transaction history
- Digital wallets looking to support peer-to-peer payments between Open Payment enabled wallets

## Installation

You can install the Open Payments package using:

```sh

> npm install @interledger/open-payments

Author: Once #1195 is out

  keyId: KEY_ID,
  privateKey: PRIVATE_KEY,
  paymentPointerUrl: PAYMENT_POINTER_URL
})
```

In order to create the client, three properties need to be provided: `keyId`, the `privateKey` and the `paymentPointerUrl`:

| Variable            | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | `paymentPointerUrl` | The valid payment pointer with which the client making requests will identify itself. A JSON Web Key Set document, including the public key that the client instance will use to protect requests and any user-facing information about the client instance used in interactions, MUST be available at the `{paymentPointerUrl}/jwks.json` url. This will be used as the `client` field during [Grant Creation](https://docs.openpayments.guide/reference/post-request). |

Reviewer 2: `any user-facing information about the client instance used in interactions` would be on the payment pointer itself (as opposed to `/jwks.json`)

> | `privateKey`        | The the private EdDSA-Ed25519 key bound to the payment pointer, and used to sign the authenticated requests with. As mentioned above, the public JWK document of this key MUST be available at the `{paymentPointerUrl}/jwks.json` url.                                                                                                                                                                                                                                  |

Reviewer 2: ```suggestion
| `privateKey`        | The private EdDSA-Ed25519 key bound to the payment pointer, and used to sign the authenticated requests with. As mentioned above, the public JWK document of this key MUST be available at the `{paymentPointerUrl}/jwks.json` url.                                                                                                                                                                                                                                  |
```

  // The public JWK with this key (and keyId) would be available at https://happy-life-bank/shoe-shop/jwks.json
})
```

2.  Get `PaymentPointers`

Grab the payment pointers of the parties:

```ts
const shoeShopPaymentPointer = await client.paymentPointer.get({
  url: 'https://happy-life-bank/shoe-shop'
})

const customerPaymentPointer = await client.paymentPointer.get({
  url: 'https://cloud-nine-wallet/alice'
})
```

3. Create `IncomingPayment`

Shoe Shop's backend gets a grant (if it doesn't have one already) to create an `IncomingPayment` on its wallet:

```ts
const incomingPaymentGrant = await client.grant.request(
  { url: shoeShopPaymentPointer.authServer },
  {
    access_token: {
      access: [
        {
          type: 'incoming-payment',
          actions: ['read-all', 'create']
        }
      ]
    }
  }
)

const incomingPayment = await client.incomingPayment.create(
  {
    paymentPointer: shoeShopPaymentPointer.authServer,
    accessToken: incomingPaymentGrant.access_token.value
  },
  {
    incomingAmount: {
      assetCode: 'USD',
      assetScale: 2,
      value: '5000'
    },
    description: 'Purchase at Shoe Shop',
    externalRef: '#INV2022-8363828'
  }
)

> ```

Reviewer 2: Should the example use a third-party client?
As is, I'm wondering why Shoe Shop has to use Open Payments to create an incoming payment for itself.

Reviewer 2: I'm curious if a bank would customers access other than via Open Payments.
For instance to create incoming payments

Author: I think Amazon Marketplace is a good example here - I'll convert to that. I was thinking like Shopify at first, but in that example, the merchant shop/wallet boundary gets blurred again.

Reviewer 2: I wonder if it'd be safer to still use a fictitious name to avoid any issues around implications that FAANGs are (already) using our technology

4. Create `Quote`

> Then, it'll gets a grant to create a `Quote` on Alice's payment pointer, which will give the amount it'll cost Alice to make the payment (with the ILP fees + her wallet's fees)

Reviewer 2: ```suggestion
Then, it'll get a grant to create a `Quote` on Alice's payment pointer, which will give the amount it'll cost Alice to make the payment (with the ILP fees + her wallet's fees)
```

```ts
const pendingOutgoingPaymentGrant = await client.grant.request(
  { url: customerPaymentPointer.authServer },
  {
    access_token: {
      access: [
        {
          type: 'outgoing-payment',
          actions: ['create', 'read'],
          identifier: customerPaymentPointer.id,
          limits: {
            sendAmount: quote.sendAmount // This limit will insure Shoe Shop would only be able to create an OutgoingPayment only as high as the quoted amount
          }
        }
      ]
    },
    interact: {
      start: ['redirect'],
      finish: {
        method: 'redirect',
        uri: 'https://shoe-shop/open-payment/complete', // this is where Alice will be redirected to after she accepts (or rejects) the request to pull money out of her account
        nonce: '456'
      }
    }
  }
)
```

6. Get user confirmation via grant & redirect

Since the `OutgoingPayment` grant is interactive (meaning, Alice needs to consent to Shoe Shop creating an `OutgoingPayment` on her account), Shoe Shop will need to redirect Alice to where the grant points to. Say the open payment request was being processed in an `/open-payment-start` route:

```ts

> app.post('/open-payment-start', (req, res) => {

Author: its tougher to explain this part with the interactivity

Author: Maybe I can just link to the grant interaction flow #1194 here

  ...
  res.send({
    redirectUrl: pendingOutgoingPaymentGrant.interact.redirect // e.g. https://cloud-nine-wallet/interact/...
  })
})
```

> From there on, the webshop can take Alice to a URL (e.g. https://cloud-nine-wallet/interact/../) where she can approve the request for Shoe Shop to create an `OutgoingPayment` on her wallet (i.e. take money out of her account). Once we completes her interaction, she should be redirected to the `interact.finish.uri` url provided in the initial grant request.

Reviewer 2: ```suggestion
From there on, the Shoe Shop website can take Alice to a URL (e.g. https://cloud-nine-wallet/interact/../) where she can approve the request for Shoe Shop to create an `OutgoingPayment` on her wallet (i.e. take money out of her account). Once she completes her interaction, she should be redirected to the `interact.finish.uri` url provided in the initial grant request.
```
